### PR TITLE
feat: added support for update noop operations

### PR DESCRIPTION
**Pull request by gpuente**
## Description:

When 2 or more UNDO operations are dispatched in a row, new NOOP operations are not considered as "new operations", instead they are considered as "update" operations (index value is not increase, but skip value is updated)

This PR adds support for that scenario. When a operation is detected as an update of a previous NOOP, the new operation is handled as an update of the latest NOOP

> NOTE: This PR requires https://github.com/powerhouse-inc/document-model/pull/30

### CODE REVIEW

File: src/server/index.ts [hunk 1]
```diff
@@ -467,18 +467,27 @@ export class DocumentDriveServer extends BaseDocumentDriveServer {
...
 
         // eslint-disable-next-line prefer-const
-        let [operationsToApply, error] = this._validateOperations(
-            operations,
-            documentStorage
-        );
+        let [operationsToApply, error, updatedOperations] =
+            this._validateOperations(operations, documentStorage);
+
+        const unregisteredOps = [
```
Comment (gpuente) [inline src/server/index.ts:478]: unregisteredOps contains information about the op and the type (new or update)

File: src/storage/browser.ts [hunk 2]
```diff
@@ -57,7 +57,8 @@ export class BrowserStorage implements IDriveStorage {
         drive: string,
         id: string,
         operations: Operation[],
-        header: DocumentHeader
+        header: DocumentHeader,
+        updatedOperations: Operation[] = []
```
Comment (gpuente) [inline src/storage/browser.ts:61]: `addDocumentOperations` now receives a new argument called `updatedOperations` that contains the operations updated.

`updatedOperations` are not included into `operations` argument

File: src/server/index.ts [hunk 3]
```diff
@@ -527,7 +550,17 @@ export class DocumentDriveServer extends BaseDocumentDriveServer {
                 .slice(0, i);
             const scopeOperations = documentStorage.operations[op.scope];
 
-            const nextIndex = scopeOperations.length + pastOperations.length;
+            // get latest operaion
```
Comment (acaldas) [inline src/server/index.ts:553]: operaion typo :p

**Review by acaldas — COMMENTED** [review]
Do we need to manually store the undone operations in the clipboard on the storage layer? Or we can build the clipboard by loading the operations?

**Comment by gpuente** [comment]
@acaldas 
> Do we need to manually store the undone operations in the clipboard on the storage layer? Or we can build the clipboard by loading the operations?

We don't have to store the skipped operations in the clipboard in this case. This is handled by the editor in the editor session. When we dispatch an undo action, the reducer takes care of transform the undo into a NOOP + skip and store the skipped op into the clipboard. In this case, when a NOOP op is received (not an UNDO action) we just have to apply the skip and not store the skipped ops into the clipboard. (clipboard is only keep in the local editor and session)

File: src/server/index.ts (hunk 3 above)
Comment (gpuente) [reply src/server/index.ts]: added fix: [84aaa77](https://github.com/powerhouse-inc/document-drive/pull/42/commits/84aaa77a86ae80dd1027a02b77d1bc19c9bd1715)

**Comment by acaldas** [comment]
> @acaldas
> 
> > Do we need to manually store the undone operations in the clipboard on the storage layer? Or we can build the clipboard by loading the operations?
> 
> We don't have to store the skipped operations in the clipboard in this case. This is handled by the editor in the editor session. When we dispatch an undo action, the reducer takes care of transform the undo into a NOOP + skip and store the skipped op into the clipboard. In this case, when a NOOP op is received (not an UNDO action) we just have to apply the skip and not store the skipped ops into the clipboard. (clipboard is only keep in the local editor and session)

But then, if for example the app is restarted, the undone operations info is lost and we no longer can redo them right?

**Pull request merged**